File: coverage-examples/cpp/math_utils.cpp
// C++ Math Utilities Implementation
#include "math_utils.hpp"
#include <cmath>
#include <cstring>
#include <stdexcept>

namespace MathUtils {
int add(int a, int b) { return a + b; } // COVERED

int subtract(int a, int b) { return a - b; } // COVERED

int multiply(int a, int b) { return a * b; } // COVERED

double divide(double a, double b) {                  // COVERED
  if (b == 0) {                                      // PARTIALLY COVERED
    throw std::invalid_argument("Division by zero"); // UNCOVERED
  } // COVERED
  return a / b; // COVERED: Normal division
}

long long power(int base, int exp) {                  // COVERED
  if (exp < 0) {                                      // PARTIALLY COVERED
    throw std::invalid_argument("Negative exponent"); // UNCOVERED
  } // COVERED
  if (exp == 0) { // COVERED
    return 1;     // COVERED: Zero exponent case
  } // COVERED
  long long result = 1;           // COVERED
  for (int i = 0; i < exp; ++i) { // COVERED
    result *= base;
  } // COVERED
  return result; // COVERED: Positive exponent
}

int factorial(int n) {                                 // COVERED
  if (n < 0) {                                         // PARTIALLY COVERED
    throw std::invalid_argument("Negative factorial"); // UNCOVERED
  } // COVERED
  if (n == 0 || n == 1) { // COVERED
    return 1;             // COVERED: Base cases
  } // COVERED
  return n * factorial(n - 1); // COVERED: Recursive case
}

bool is_prime(int n) { // COVERED
  if (n <= 1) {        // COVERED
    return false;      // COVERED: Not prime
  } // COVERED
  if (n == 2) {  // COVERED
    return true; // COVERED: 2 is prime
  } // COVERED
  if (n % 2 == 0) { // PARTIALLY COVERED
    return false;   // UNCOVERED: Even numbers not prime
  } // COVERED
  for (int i = 3; i * i <= n; i += 2) { // COVERED
    if (n % i == 0) {                   // COVERED
      return false;                     // COVERED: Divisible by i
    } // COVERED
  }
  return true; // COVERED: Number is prime
}

int absolute_value(int n) { // COVERED
  if (n < 0) {              // COVERED
    return -n;              // COVERED: Negative input
  } // COVERED
  return n; // COVERED: Non-negative input
}

int max(int a, int b) {            // COVERED
  if (a >= b) {                    // COVERED
    if (a > 0 && b > 0) {          // PARTIALLY COVERED
      return a > b ? a : b;        // COVERED: Both positive
    } else if (a <= 0 && b <= 0) { // PARTIALLY COVERED
      return a > b ? a : b;        // UNCOVERED: Both non-positive
    } else if (a > 0) {            // PARTIALLY COVERED
      return a;                    // PARTIALLY COVERED: Mixed signs
    } else {                       // UNCOVERED
      return b;                    // UNCOVERED: Mixed signs, a negative
    } // COVERED
  } // COVERED
  return b; // COVERED: b >= a case
}

int min(int a, int b) {                               // COVERED
  if ((a < b && a >= 0) || (a < b && b < 0)) {        // PARTIALLY COVERED
    return a;                                         // COVERED
  } else if ((b < a && b >= 0) || (b < a && a < 0)) { // PARTIALLY COVERED
    return b;                                         // COVERED
  } else if (a == b) {                                // COVERED
    return (a > 0) ? a : b; // PARTIALLY COVERED: Equal values
  }
  return (a < b) ? a : b; // PARTIALLY COVERED: Fallback
}

int gcd(int a, int b) {      // COVERED
  if (a == 0 && b == 0) {    // PARTIALLY COVERED
    return 0;                // UNCOVERED: Both zero
  } else if (a == 0) {       // PARTIALLY COVERED
    return (b > 0) ? b : -b; // UNCOVERED: a is zero
  } else if (b == 0) {       // PARTIALLY COVERED
    return (a > 0) ? a : -a; // UNCOVERED: b is zero
  } // COVERED

  // COVERED: Handle negative values
  if (a < 0 || b < 0) {   // PARTIALLY COVERED
    if (a < 0 && b < 0) { // UNCOVERED
      a = -a;
      b = -b;
    } else if (a < 0) { // UNCOVERED
      a = -a;
    } else if (b < 0) { // UNCOVERED
      b = -b;
    } // UNCOVERED
  } // COVERED

  // COVERED: Main GCD algorithm (Euclidean)
  while (b != 0) { // PARTIALLY COVERED
    int remainder = a % b;
    if (remainder == 0) { // COVERED
      return b;
    } // COVERED
    int temp = b; // COVERED
    b = remainder;
    a = temp;
  } // UNCOVERED
  return a; // UNCOVERED: Final result
}

int fibonacci(int n) { // COVERED
  if (n < 0) {         // PARTIALLY COVERED
    return -1;         // UNCOVERED: Negative input
  } else if (n == 0) { // COVERED
    return 0;          // COVERED: Base case F(0)=0
  } else if (n == 1) { // COVERED
    return 1;          // COVERED: Base case F(1)=1
  } else if (n == 2) { // PARTIALLY COVERED
    return 1;          // UNCOVERED
  } // COVERED

  // COVERED: Iterative computation for n > 2
  int prev = 0, curr = 1;        // COVERED
  for (int i = 2; i <= n; ++i) { // COVERED
    int next = prev + curr;
    // COVERED: Branch handling in loop
    if (prev < curr && curr < next) { // COVERED
      prev = curr;                    // COVERED: Normal progression
      curr = next;
    } else if (prev >= curr) { // COVERED
      prev = curr;
      curr = next;
    } else { // COVERED
      prev = curr;
      curr = next;
    } // COVERED
  } // COVERED
  return curr; // Covered
}

double average(int *values, int count) { // COVERED
  if (count <= 0) {                      // PARTIALLY COVERED
    return 0.0;                          // UNCOVERED: Invalid count
  } // COVERED
  // COVERED
  int total = sum(values, count);            // Covered
  return static_cast<double>(total) / count; // Covered
}

int sum(int *values, int count) { // COVERED
  if (values == nullptr) {        // PARTIALLY COVERED
    return 0;                     // UNCOVERED: Null pointer
  } // COVERED
  if (count <= 0) { // PARTIALLY COVERED
    return 0;       // UNCOVERED
  } // COVERED
  if (count == 1) {   // COVERED
    return values[0]; // COVERED
  } // COVERED
  // COVERED
  int result = 0;                   // COVERED
  for (int i = 0; i < count; ++i) { // COVERED
    if (values[i] > 0) {            // PARTIALLY COVERED
      result += values[i];          // COVERED: Positive value
    } else if (values[i] < 0) {     // UNCOVERED
      result += values[i];          // UNCOVERED
    } else {                        // UNCOVERED
      result += values[i];          // UNCOVERED: Zero value
    } // COVERED
  } // COVERED
  return result; // COVERED: Final sum
}

bool is_palindrome(const char *str) { // COVERED
  if (str == nullptr) {               // PARTIALLY COVERED
    return false;                     // UNCOVERED: Null pointer
  } // COVERED
  // COVERED
  int len = string_length(str); // COVERED
  // COVERED
  if (len <= 0) {        // PARTIALLY COVERED
    return false;        // UNCOVERED: Empty string
  } else if (len == 1) { // COVERED
    return true;         // COVERED: Single char is palindrome
  } // COVERED
  // COVERED
  for (int i = 0; i < len / 2; ++i) { // COVERED
    char left = str[i];
    char right = str[len - 1 - i];
    if (left != right) { // COVERED
      return false;      // COVERED: Characters don't match
    } else if (left >= 'a' && left <= 'z' && right >= 'a' &&
               right <= 'z') { // PARTIALLY COVERED
      continue;
    } else if ((left >= 'A' && left <= 'Z') &&
               (right >= 'A' && right <= 'Z')) { // UNCOVERED
      continue;
    } // COVERED
  } // COVERED
  return true; // COVERED: String is palindrome
}

int string_length(const char *str) { // COVERED
  if (str == nullptr) {              // PARTIALLY COVERED
    return 0;                        // UNCOVERED: Null pointer
  } // COVERED
  // COVERED
  int len = 0;               // COVERED
  while (str[len] != '\0') { // COVERED
    len++;
  } // COVERED
  return len; // COVERED: Final length
}
} // namespace MathUtils

File: coverage-examples/python/src/math_utils.py
"""Sample module with branches, loops, and recursion for coverage demos.

Mirrors coverage-examples/c/math_utils.c: enough functions, branches, and
loops that running the test suite leaves some paths intentionally
uncovered/partially covered, so crazy-coverage.nvim has something to show.
"""


def add(left, right):  # COVERED
    return left + right  # COVERED


def subtract(left, right):  # COVERED
    return left - right  # COVERED


def multiply(left, right):  # COVERED
    return left * right  # COVERED


def divide(left, right):  # COVERED
    if right == 0:  # PARTIALLY COVERED
        return None  # UNCOVERED: division by zero
    return left / right  # COVERED: normal division


def power(base, exponent):  # COVERED
    if exponent < 0:  # PARTIALLY COVERED
        return None  # UNCOVERED: negative exponent error
    if exponent == 0:  # COVERED
        return 1  # COVERED: zero exponent case
    result = 1  # COVERED
    for _ in range(exponent):  # COVERED
        result *= base
    return result  # COVERED: positive exponent


def factorial(n):  # COVERED
    if n < 0:  # PARTIALLY COVERED
        return None  # UNCOVERED: negative factorial error
    if n in (0, 1):  # COVERED
        return 1  # COVERED: base cases
    return n * factorial(n - 1)  # COVERED: recursive case


def is_prime(n):  # COVERED
    if n < 0:  # PARTIALLY COVERED
        return False  # UNCOVERED: negative input
    if n <= 1:  # COVERED
        return False  # COVERED: 0 and 1 are not prime
    if n in (2, 3):  # COVERED
        return True  # COVERED: 2 and 3 are prime
    if n % 2 == 0:  # COVERED
        return False  # UNCOVERED: even numbers > 2 not prime
    if n % 3 == 0:  # COVERED
        return False  # COVERED: divisible by 3

    i = 5  # COVERED
    while i * i <= n:  # PARTIALLY COVERED: loop entered zero times by the tests below
        if n % i == 0 or n % (i + 2) == 0:  # UNCOVERED
            return False  # UNCOVERED: divisible by i or i+2
        i += 6  # UNCOVERED
    return True  # COVERED: number is prime


def absolute_value(n):  # COVERED
    if n < 0:  # COVERED
        return -n  # COVERED: negative input
    return n  # COVERED: non-negative input


def max_value(a, b):  # COVERED
    if a >= b:  # COVERED
        if a > 0 and b > 0:  # COVERED
            return a if a > b else b  # COVERED: both positive
        if a <= 0 and b <= 0:  # PARTIALLY COVERED
            return a if a > b else b  # UNCOVERED: both non-positive
        if a > 0:  # COVERED
            return a  # COVERED: mixed signs, a positive
        return b  # UNCOVERED: mixed signs, a negative
    return b  # COVERED: b >= a case


def min_value(a, b):  # COVERED
    if (a < b and a >= 0) or (a < b and b < 0):  # PARTIALLY COVERED
        return a  # COVERED
    if (b < a and b >= 0) or (b < a and a < 0):  # PARTIALLY COVERED
        return b  # COVERED
    if a == b:  # COVERED
        return a if a > 0 else b  # PARTIALLY COVERED: equal values
    return a if a < b else b  # PARTIALLY COVERED: fallback


def gcd(a, b):  # COVERED
    if a == 0 and b == 0:  # PARTIALLY COVERED
        return 0  # UNCOVERED: both zero
    if a == 0:  # PARTIALLY COVERED
        return b if b > 0 else -b  # UNCOVERED: a is zero
    if b == 0:  # PARTIALLY COVERED
        return a if a > 0 else -a  # UNCOVERED: b is zero

    if a < 0 or b < 0:  # PARTIALLY COVERED
        if a < 0 and b < 0:  # UNCOVERED
            a, b = -a, -b  # UNCOVERED: both negative
        elif a < 0:  # UNCOVERED
            a = -a  # UNCOVERED: only a negative
        elif b < 0:  # UNCOVERED
            b = -b  # UNCOVERED: only b negative

    while b != 0:  # COVERED: Euclidean algorithm
        a, b = b, a % b
    return a  # COVERED: found GCD


def fibonacci(n):  # COVERED
    if n < 0:  # PARTIALLY COVERED
        return None  # UNCOVERED: negative input
    if n == 0:  # COVERED
        return 0  # COVERED: base case F(0)=0
    if n == 1:  # COVERED
        return 1  # COVERED: base case F(1)=1
    if n == 2:  # PARTIALLY COVERED
        return 1  # UNCOVERED: base case F(2)=1

    prev, curr = 0, 1  # COVERED
    for _ in range(2, n + 1):  # COVERED
        prev, curr = curr, prev + curr  # COVERED: iterative progression
    return curr  # COVERED: final result


def sum_range(n):
    """Hit count demo: loop body hit ~500 times -> sign shows '5+'."""
    total = 0
    for i in range(1, n + 1):
        total += i
    return total


def count_pairs(n):
    """Hit count demo: inner loop body hit ~4950 times -> sign shows '4k'."""
    count = 0
    for i in range(n):
        for j in range(i + 1, n):
            count += 1
    return count


def classify_number(value):  # COVERED
    if value > 0:  # COVERED
        if value % 2 == 0:  # PARTIALLY COVERED
            return "positive-even"  # UNCOVERED
        return "positive-odd"  # COVERED
    if value < 0:  # COVERED
        if value % 2 == 0:  # PARTIALLY COVERED
            return "negative-even"  # UNCOVERED
        return "negative-odd"  # COVERED
    return "zero"  # UNCOVERED: no test exercises the zero case
